From d23591bc4306e607c260d7ab7873a0e905e4479b Mon Sep 17 00:00:00 2001
From: Andrey Ulanov <andreyu@google.com>
Date: Wed, 16 Nov 2016 16:12:51 -0800
Subject: [PATCH] When creating users try to derive the UID/GID from the
 homedir

It's possible that when a new user is being created that the home
directory for that user already exists. When this happens account
manager will now try to find the UID of the user from that directory
and reuse it.

This fixes potential problems if for some reason the new accounts that
were previously added into passwd/group were not preserved after reboot,
but the correspondinghome directory was.
---
 google_compute_engine/accounts/accounts_utils.py | 35 +++++++++++++++++++++---
 1 file changed, 31 insertions(+), 4 deletions(-)

diff --git a/google_compute_engine/accounts/accounts_utils.py b/google_compute_engine/accounts/accounts_utils.py
index fbdd5ef..5c90226 100644
--- a/google_compute_engine/accounts/accounts_utils.py
+++ b/google_compute_engine/accounts/accounts_utils.py
@@ -104,11 +104,13 @@ class AccountsUtils(object):
     except KeyError:
       return None
 
-  def _AddUser(self, user):
+  def _AddUser(self, user, uid=None, homedir=None):
     """Configure a Linux user account.
 
     Args:
       user: string, the name of the Linux user account to create.
+      uid: string, derived from 'new' user's home directory
+      homedir: string, home directory for 'new' user
 
     Returns:
       bool, True if user creation succeeded.
@@ -126,7 +128,14 @@ class AccountsUtils(object):
     #
     # To solve the issue, make the password '*' which is also recognized
     # as locked but does not prevent SSH login.
-    command = ['useradd', '-m', '-s', '/bin/bash', '-p', '*', user]
+    if homedir:
+      extra_flags = ['-d', homedir]
+    else:
+      extra_flags = ['-m']
+    if uid:
+      extra_flags += ['-u', str(uid)]
+
+    command = ['useradd', user, '-s', '/bin/bash', '-p', '*'] + extra_flags
     try:
       subprocess.check_call(command)
     except subprocess.CalledProcessError as e:
@@ -284,11 +293,29 @@ class AccountsUtils(object):
     if not bool(USER_REGEX.match(user)):
       self.logger.warning('Invalid user account name %s.', user)
       return False
+
     if not self._GetUser(user):
       # User does not exist. Attempt to create the user and add them to the
       # appropriate user groups.
-      if not (self._AddUser(user) and
-              self._UpdateUserGroups(user, self.groups)):
+      user_created = False
+
+      # Check if the homedir already exists and if yes try to derive UID
+      # and GID from that directory.
+      homedir = '/home/%s' % user
+      if os.path.isdir(homedir):
+        stats = os.stat(homedir)
+        if stats.st_uid > 1000 and stats.st_gid == stats.st_uid:
+          self.logger.info('Using UID %d and homedir %s for account %s',
+                       stats.st_uid, homedir, user)
+
+          if not self._AddUser(user, uid=stats.st_uid, homedir=homedir):
+            return False
+          user_created = True
+
+      if not user_created and not self._AddUser(user):
+        return False
+
+      if not self._UpdateUserGroups(user, self.groups):
         return False
 
     # Don't try to manage account SSH keys with a shell set to disable
-- 
2.8.0.rc3.226.g39d4020

